doc/man/wgetigsrnx.1
.\" Man page generated from reStructuredText.
.
.TH "WGETIGSRNX" "1" "October 2014" "AutoBpe" "User Commands"
.SH NAME
wgetigsrnx \- AutoBpe Software Collection
.SH INTRODUCTION
.sp
AutoBpe Software Collection is a collection of tools that accomodates
automatic processing of GNSS data mainly via Bernese GNSS.
.SH SYNOPSIS
.B wgetigsrnx
[\fIOPTIONS\fR]... \fI\-y [YEAR]\fR \fI\-d [DAY OF YEAR]\fR
.SH DESCRIPTION
.\" Add any additional description here
.PP
Download RINEX files (Hatanaka compressed) for IGS stations for a specific date. The data centers used in this script to
download the files, are 
.IP \[bu] 2
ftp://cddis.gsfc.nasa.gov/pub/gps/data/daily and 
.IP \[bu]
ftp://igs.bkg.bund.de/IGS/obs/
.PP
Mandatory arguments for long options are mandatory for short options too.
.TP
\fB\-s\fR, \fB\-\-stations\fR
specify stations to download seperated by whitespace (e.g -s penc pdel ...), see \fBSTATION PRIORITY\fR
.TP
\fB\-f\fR, \fB\-\-station-file\fR
give a file specifying stations to be downloaded. The file must have one line, where each station is seperated
by a whitespace character, see \fBSTATION PRIORITY\fR
.TP
\fB\-y\fR, \fB\-\-year\fR
specify year (4-digits)
.TP
\fB\-d\fR, \fB\-\-doy\fR
specify day of year
.TP
\fB\-o\fR, \fB\-\-output-directory\fR
specify output directory where the downloaded files are going to be saved (default ./)
.TP
\fB\-m\fR, \fB\-\-max-stations\fR
specify a station limit. When this limit is reached, no more stations are going to be
downloaded.
.TP
\fB\-r\fR, \fB\-\-force-remove\fR
if a file is found with the same name as the one to be downloaded, it will be deleted before downloading.
The default behaviour is that if a file is found with the same name as the one to be downloaded, then
the file will \fBNOT\fR be downloaded but counted as downloaded.
.TP
\fB\-z\fR, \fB\-\-decompress-Z\fR
decompress the downloaded files (default do not decompress). If this switch is not enabled and a file
is found matching a file to be downloaded but not compressed, then the existing file is going to be
compressed (to match exactly the file as it would be downloaded).
.TP
\fB\-n\fR, \fB\-\-upper-case\fR
truncate all downloaded files to upper case letters.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
supress all messages, except for errors.
.TP
\fB\-h\fR, \fB\-\-help\fR
only print the help message and exit
\fB\-v\fR, \fB\-\-version\fR
only print version message and exit
.SH "EXIT STATUS"
The program will return an integer, denoting the exit status. There are three distinct cases,
.IP \[bu] 2
Exit Status = 255-2; user has not provided any stations to download
.IP \[bu] 2
Exit Status = 255-1; this denotes that an error has occured
.IP \[bu] 2
Exit Status >=0 ; this denotes that the program had a seccesseful run and downloaded this many station files.
Note that some of the stations may have not actualy been downloaded by this run, but the wanted file(s) was
found.
.SH "STATION PRIORITY"
In case both -s and -f switches are used, the order of the stations is first 
the -s stations and then the -f ones (e.g > wgetigsrnx -s sta1 sta2 -f filename ...
and filename has stations sta3 sta4, then the list of stations to download will be:
sta1 sta2 sta3 sta4). The order of stations is crucial when the -m switch is specified.
.SH "REPORTING BUGS"
Report wgetigsrnx bugs to 
.br
Demitris Anastasiou <danast@mail.ntua.gr>
.br
Xanthos Papanikolaou <xanthos@mail.ntua.gr>
.SH "AUTHOR"
National Technical University of Athens,
.br
School of Rural and Surveying Engineering,
.br
Dionysos Satellite Observatory, Higher Geodesy Laboratory
.br


bash/wgetigsrnx.sh
#! /bin/bash

################################################################################
## 
## |===========================================|
## |** Higher Geodesy Laboratory             **|
## |** Dionysos Satellite Observatory        **| 
## |** National Tecnical University of Athens**|
## |===========================================|
##
## filename              : wgetigsrnx.sh
                           NAME=wgetigsrnx
## version               : v-1.0
                           VERSION=v-1.0
                           RELEASE=beta
## created               : APR-2013
## usage                 : wgetigsrnx.sh
## exit code(s)          : >=0 -> success (actual number of stations downloaded)
##                       :  -1 -> error
##                       :  -2 -> error (no stations to download)
## discription           : downloads igs rinex files.
## uses                  : wget
## notes                 : this is a newer version of the script downloadigs.sh
##                         Updated lists of IGS stations can be found at:
##                         ftp://igscb.jpl.nasa.gov/pub/station
## TODO                  :
## detailed update list  : DEC-2013 corrected exit codes
##                       : OCT-2014
                           LAST_UPDATE=OCT-2014
##
################################################################################

# //////////////////////////////////////////////////////////////////////////////
# DISPLAY VERSION
# //////////////////////////////////////////////////////////////////////////////
function dversion {
  echo "${NAME} ${VERSION} (${RELEASE}) ${LAST_UPDATE}"
  exit 0
}

# //////////////////////////////////////////////////////////////////////////////
# HELP FUNCTION
# //////////////////////////////////////////////////////////////////////////////
function help {
  echo "/******************************************************************************************/"
  echo " Program Name : wgetigsrnx.sh"
  echo " Version      : $VERSION"
  echo " Last Update  : $LAST_UPDATE"
  echo ""
  echo " Purpose : Download IGS stations rinex files for a specific date. If the file already exists,"
  echo "           (in capital letters, compressed or not, Hatanaka or RINEX) then it will not "
  echo "           be downloaded, except if the -r switch is given. In case a file already exists and"
  echo "           does not to be downloaded, this file is counted as a downloaded one in the integer returned."
  echo ""
  echo " Usage   : wgetigsrnx.sh [-s [station1] [station2] [...] | -f [station file]] -y [year] -d [doy] {-o [output directory] | -m [max stations] | -r}"
  echo ""
  echo " Switches: "
  echo "           -s --stations specify stations to download seperated by "
  echo "            whitespace (e.g -s penc pdel ...) (*)"
  echo "           -f --station-file give a file specifying stations to be downloaded"
  echo "            File must have one line, where each station is seperated"
  echo "            by a whitespace character.(*)"
  echo "           -y --year 4-digit year of date to download (e.g. -y 2010)"
  echo "           -d --doy day of year [1-366] to download (e.g. -d 35)"
  echo "           -o --output-directory directory to save rinex files. Can be"
  echo "            either relative or absolute path not followed by / (e.g. /home/bpe/data)"
  echo "           -m --max-stations total number of allowed stations. Only a total"
  echo "            of max_stations will be downloaded. If more stations are specified"
  echo "            they will be skipped if max_stations is reached (e.g. -m 15)"
  echo "           -r --force-remove if a file is found with the same name as the one"
  echo "            to be downloaded, it will be deleted before downloading."
  echo "           -h --help display (this) help message and exit"
  echo "           -z --decompress decompress the downloaded rinex files"
  echo "           -n --upper-case rename (truncate) all downloaded files to capital letters"
  echo "           -q --quiet supress all mesagges"
  echo " [OBSOLETE]-t [:= file size] if -r switch is not given, then remove all files same as the"
  echo "            ones to be downloaded if their size is smaller than <file size> in bytes"
  echo "           -v --version dsiplay version and exit"
  echo ""
  echo " Exit Status:255-2 -> no stations to download provided"
  echo " Exit Status:255-1 -> error"
  echo " Exit Status: >= 0 -> sucesseful exit (actual number of stations downloaded if # stations < 255)"
  echo ""
  echo "(*) In case both -s and -f switches are used, the order of the stations is first"
  echo "the -s stations and then the -f ones (e.g > wgetigsrnx -s sta1 sta2 -f filename ..."
  echo "and filename has stations sta3 sta4, then the list of stations to download will be:"
  echo "sta1 sta2 sta3 sta4). The order of stations is crucial when the -m switch is specified."
  echo ""
  echo " |===========================================|"
  echo " |** Higher Geodesy Laboratory             **|"
  echo " |** Dionysos Satellite Observatory        **|"
  echo " |** National Tecnical University of Athens**|"
  echo " |===========================================|"
  echo ""
  echo "/******************************************************************************************/"
  exit 0
}

# //////////////////////////////////////////////////////////////////////////////
# PRE-DEFINE BASIC VARIABLES
# //////////////////////////////////////////////////////////////////////////////
YEAR=-1900    ## year
YR2=00        ## last two digits of year
DOY=0         ## day of year
STARRAY=()    ## array of igs stations
OUTDIR=$(pwd) ## directory to save rinex files
IGSURL1=ftp://cddis.gsfc.nasa.gov/pub/gps/data/daily ## url for igs rinex files
IGSURL2=ftp://igs.bkg.bund.de/IGS/obs                ## url for igs rinex files
NUMOFTRIES=3  ## number of tries for wget
STACOUNTER=0  ## number of stations actualy downloaded
DOWNLOADED=0  ## flag to mark if a specific rinex has been downloaded
NUMOFST=0     ## total number of stations to download
MAXST=0       ## max number of stations to download (all stations after this 
              ## number is reached will not be downloaded)
FORCE=0       ## delete any other rinex file with the same name already in the dir
STFILE=0      ## filename of the station file
DECOMP=0      ## decompress the rinex files
TRUNCATE=NO   ## truncate to upper case
QUIET=NO      ## supress messages

# //////////////////////////////////////////////////////////////////////////////
# GET COMMAND LINE ARGUMENTS
# //////////////////////////////////////////////////////////////////////////////
if [ "$#" == "0" ]
then
  help
  exit 254
fi

while [ $# -gt 0 ]
do
  case "$1" in
    -s|--stations)
      shift
      while [ $# -gt 0 ]
      do
        j=$1
        if [ ${j:0:1} == "-" ]
        then
          break
        else
          STARRAY+=($j)
          shift
        fi
      done
      ;;
    -y|--year)
      YEAR=${2}
      shift
      shift
      ;;
    -d|--doy)
      DOY=`echo $2 | sed 's/^0*//'`
      shift
      shift
      ;;
    -o|--output-directory)
      OUTDIR=${2}
      shift
      shift
      ;;
    -m|--max-stations)
      MAXST=${2}
      shift
      shift
      ;;
    -r|--force-remove)
      FORCE=1
      shift
      ;;
    -h|--help)
      help
      ;;
    -f|--station-file)
      STFILE=${2}
      shift
      shift
      ;;
    -z|--decompress)
      DECOMP=1
      shift
      ;;
    -t|--max-size)
      if [ $QUIET == "NO" ]; then echo "WARNING! -t switch is obsolete and will be ignored."; fi
      shift
      shift
      ;;
    -n|--upper-case)
      TRUNCATE=YES
      shift
      ;;
    -q|--quiet)
      QUIET=YES
      shift
      ;;
    -v|--version)
      dversion
      ;;
  esac
done

# //////////////////////////////////////////////////////////////////////////////
# IF FILE IS GIVEN, ADD ITS ENTRIES TO THE STATION LIST
# //////////////////////////////////////////////////////////////////////////////
if [ "$STFILE" != "0" ]; then
  if [ -f "$STFILE" ]; then
    TEMPARRAY=( $( cat "$STFILE" ) )
    for j in ${TEMPARRAY[*]}; do
      STARRAY+=($j)
    done
  else
    if [ $QUIET == "NO" ]; then echo "*** File ${STFILE} does not exist!"; fi
  fi
fi

# //////////////////////////////////////////////////////////////////////////////
# CHECK THAT STATION LIST IS NOT EMPTY
# //////////////////////////////////////////////////////////////////////////////
if [ ${#STARRAY[*]} -eq 0 ]; then
  if [ $QUIET == "NO" ]; then echo "*** Need to provide at least one IGS station"; fi
  exit 253
fi
NUMOFST=${#STARRAY[*]}

# //////////////////////////////////////////////////////////////////////////////
# CHECK IF MAX STATIONS IS SET, ELSE SET IT TO NUMBER OF STATIONS IN LIST
# //////////////////////////////////////////////////////////////////////////////
if [ "$MAXST" -eq 0 ]; then
  MAXST=$NUMOFST
fi

# //////////////////////////////////////////////////////////////////////////////
# CHECK THAT YEAR AND DOY IS SET AND VALID
# //////////////////////////////////////////////////////////////////////////////
if [ $YEAR -lt 1950 ]; then
  if [ $QUIET == "NO" ]; then echo "*** Need to provide a valid year [>1950]"; fi
  exit 254
fi
YR2=${YEAR:2:2}
if [ $DOY -lt 1 ] || [ $DOY -gt 366 ]; then
  if [ $QUIET == "NO" ]; then echo "*** Need to provide a valid doy [1-366]"; fi
  exit 254
fi

# //////////////////////////////////////////////////////////////////////////////
# IF OUTPUT DIR GIVEN, SEE THAT IT EXISTS
# //////////////////////////////////////////////////////////////////////////////
if [ ! -d $OUTDIR ]; then
  if [ $QUIET == "NO" ]; then echo "*** Directory $OUTDIR does not exist!"; fi
  exit 254
fi

# //////////////////////////////////////////////////////////////////////////////
# IF DECOMPRESS IS CALLED, CHECK THAT uncompress EXISTS
# //////////////////////////////////////////////////////////////////////////////
if [ "$DECOMP" -eq 1 ]; then
  hash uncompress 2>/dev/null
  if [ "$?" -ne 0 ]; then
    if [ $QUIET == "NO" ]; then echo "*** Asked for decompress but programm <uncompress> is not available"; fi
    if [ $QUIET == "NO" ]; then echo "*** Rinex files will not be decompressed"; fi
    DECOMP=0
  fi
fi

# //////////////////////////////////////////////////////////////////////////////
# MAKE DOY A 3-DIGIT NUMBER
# //////////////////////////////////////////////////////////////////////////////
if [ $DOY -lt 10 ]; 
  then DOY=00${DOY}
elif [ $DOY -lt 100 ]
  then DOY=0${DOY}
else
  DOY=${DOY}
fi

# //////////////////////////////////////////////////////////////////////////////
# DOWNLOAD EVERY IGS STATION IN LIST
# //////////////////////////////////////////////////////////////////////////////
for station in ${STARRAY[*]}; do

  # station in capital letters
  STATION=`echo $station | tr 'a-z' 'A-Z'`

  # file to save with no ending char(s)
  f2s=${OUTDIR}/${station}${DOY}0.${YR2}
  F2S=${OUTDIR}/${STATION}${DOY}0.${YR2}

  # only try to download if max number of stations not reached
  if [ "$STACOUNTER" -lt "$MAXST" ]; then

    # if flag is set, force delete any previous rinex with same name
    if [ "$FORCE" -eq 1 ]; then
      if [ -f ${f2s}o ]; then #check and remove rinex if it exists
        rm -f ${f2s}o
      fi
      if [ -f ${f2s}d ]; then #check and remove Hatanaka if it exists
        rm -f ${f2s}d
      fi
      if [ "$DECOMP" -eq 1 ]; then #check and remove Hatanaka compressed if it exists
        if [ -f ${f2s}d.Z ]; then
          rm -f ${f}
        fi
      fi
    fi ##if [ "$FORCE" -eq 1 ]

    PROCED_TO_DOWNLOAD=YES ## if the file already exists, this will be set to "NO"
    # check to see if the file already exists, either as Hatanaka (compressed or not) or as rinex
    if [ -f ${f2s}d.Z ]; then # the compressed Hatanaka version exists
      if [ $QUIET == "NO" ]; then echo "File ${f2s}d.Z exists as Hatanaka compressed"; fi
      if [ "$DECOMP" -eq 1 ]; then
        uncompress ${f2s}d.Z
      fi
      PROCED_TO_DOWNLOAD=NO
    fi
    if [ -f ${f2s}d ]; then # the uncompressed Hatanaka version exists
      if [ $QUIET == "NO" ]; then echo "File ${f2s}d exists as Hatanaka uncompressed"; fi
      if [ "$DECOMP" -ne 1 ]; then
        compress ${f2s}d
      fi
      PROCED_TO_DOWNLOAD=NO
    fi
    if [ -f ${f2s}o ]; then # the rinex version exists
      if [ $QUIET == "NO" ]; then echo "File ${f2s}o exists as rinex"; fi
      PROCED_TO_DOWNLOAD=NO
    fi
    # check to see if the file already exists, either as Hatanaka (compressed or not) or as rinex UPPER CASE
    if [[ $TRUNCATE == "YES" && $PROCED_TO_DOWNLOAD == "YES" ]]; then
      if [ -f ${F2S}D.Z ]; then # the compressed Hatanaka version exists
        if [ $QUIET == "NO" ]; then echo "File ${F2S}D.Z exists as Hatanaka compressed"; fi
        if [ "$DECOMP" -eq 1 ]; then
          uncompress ${F2S}D.Z
        fi
        PROCED_TO_DOWNLOAD=NO
      fi
      if [ -f ${F2S}D ]; then # the uncompressed Hatanaka version exists
        if [ $QUIET == "NO" ]; then echo "File ${F2S}D exists as Hatanaka uncompressed"; fi
        if [ "$DECOMP" -ne 1 ]; then
          compress ${F2S}D
        fi
        PROCED_TO_DOWNLOAD=NO
      fi
      if [ -f ${F2S}O ]; then # the rinex version exists
        if [ $QUIET == "NO" ]; then echo "File ${F2S}O exists as rinex"; fi
        PROCED_TO_DOWNLOAD=NO
      fi
    fi ##if [ $TRUNCATE == "YES" ]

    if [ ${PROCED_TO_DOWNLOAD} == "NO" ]; then ## do not proceed to download; file available
      let STACOUNTER=STACOUNTER+1

    else ## preceed to download
      # assume file is not downloaded
      DOWNLOADED=0

      # try downloading from cddis
      wget -nc --tries="$NUMOFTRIES" -q -O ${OUTDIR}/${station}${DOY}0.${YR2}d.Z \
        ${IGSURL1}/${YEAR}/${DOY}/${YR2}d/${station}${DOY}0.${YR2}d.Z
      # if station not downloaded, try BKG url
      if [ "$?" -ne 0 ]; then
        if [ $QUIET == "NO" ]; then echo "## Station $station not found in cddis; trying in bkg ..."; fi
        rm -f ${OUTDIR}/${station}${DOY}0.${YR2}d.Z 2>/dev/null ## remove empty file if any
        wget -nc --tries="$NUMOFTRIES" -q -O ${OUTDIR}/${station}${DOY}0.${YR2}d.Z \
          ${IGSURL2}/${YEAR}/${DOY}/${station}${DOY}0.${YR2}d.Z
        # check if it is downloaded from bkg
        if [ "$?" -ne 0 ]; then
          if [ $QUIET == "NO" ]; then echo "*** Station $station not found!"; fi
          rm -f ${OUTDIR}/${station}${DOY}0.${YR2}d.Z 2>/dev/null ## remove empty file if any
        else
          DOWNLOADED=1
        fi
      else
        DOWNLOADED=1
      fi ## if [ "$?" -ne 0 ]

      # if station downloaded, augment counter
      if [ "$DOWNLOADED" -eq 1 ]; then
        let STACOUNTER=STACOUNTER+1
        if [ $QUIET == "NO" ]; then echo "Station $station downloaded"; fi
          
        # if decompressed called, uncompress the file
        if [ "$DECOMP" -eq 1 ]; then
          uncompress ${OUTDIR}/${station}${DOY}0.${YR2}d.Z
        fi

        # truncate name if needed
        if [ $TRUNCATE == "YES" ]; then
          mv ${OUTDIR}/${station}${DOY}0.${YR2}d ${OUTDIR}/${STATION}${DOY}0.${YR2}D 2>/dev/null
          mv ${OUTDIR}/${station}${DOY}0.${YR2}d.Z ${OUTDIR}/${STATION}${DOY}0.${YR2}D.Z 2>/dev/null
        fi

      fi ## if [ "$DOWNLOADED" -eq 1 ]

    fi ##if [ ${PROCED_TO_DOWNLOAD} == "NO" ]

  else ##if [ "$STACOUNTER" -lt "$MAXST" ]
    if [ $QUIET == "NO" ]; then echo "Max number of stations reached! Skipping station $station"; fi

  fi ##if [ "$STACOUNTER" -lt "$MAXST" ]

done ##for station in ${STARRAY[*]}

# //////////////////////////////////////////////////////////////////////////////
# PRINT NUMBER OF STATIONS DOWNLOADED
# //////////////////////////////////////////////////////////////////////////////
if [ $QUIET == "NO" ]; then echo "Downloaded $STACOUNTER / $NUMOFST stations"; fi

# //////////////////////////////////////////////////////////////////////////////
# SUCESEFUL EXIT
# //////////////////////////////////////////////////////////////////////////////
exit $STACOUNTER;
